### ISSUE TITLE
Impossible cave

### ISSUE THREAD

**Issue by makc** [felixmariotto/Edelweiss#21]
Ok I said I would not be bothering you this much but... here we go again :) Check this out:

https://www.youtube.com/watch?v=l9caDly8tN8

We enter the cave like 10 meters to the right of the stream, go to the right, then (almost) straight down, then exit to the right and end up like 10 meters to the left of the stream. This should not be possible, or - if this cave is supposed to go under the stream, then significant portion of it has to go to the left.

(just in case, the character disappearing when you enter the cave is also the canary-only issue, you can ignore it)

**Comment by felixmariotto** [comment]
Ho yes I know that's it's physically impossible, and this one cave is nothing compared to the caves further up the mountain, like the one behind the big pillar. All the caves are intended to make the experience of back-tracking less tedious when you need to look for more edelweiss in order to go on with the game : 
- firstly you don't have to go through the same map, so it's more interesting
- secondly, as the camera does not turn backward, some platform stages of the main map can be hard and frustrating when done in the reverse direction

That's why I came up with the idea of the caves, it's like a more interesting teleportation to backtrack, but I didn't want the player to spend too much time running toward the camera, so I made them short.

If you have an idea for a nicer cave though go ahead, I will probably merge since it only involves updating the jsons and glbs.

**Comment by makc** [comment]
> some platform stages of the main map can be hard and frustrating when done in the reverse direction

could be solved with these things

![](https://media-cdn.tripadvisor.com/media/photo-s/08/5e/a5/d2/lago-di-molveno.jpg)

if you can find good excuse why they do not work on your way up. like having to pay for this and not having money, Idk. but that is the task for level 2, if ever.

**Comment by felixmariotto** [comment]
Yes that's a good idea.   
Did you play Steamworld Dig 2 ? There is a similar concept. In this case the teleportation machine is a tube system to go underground, but the entrances are stuck with rubble, and when you get to them from the outside you can remove the rubble and use the teleportation in both directions. In case of the ski lift, maybe they could need reparation from the player ? Which would create the reason to climb another mountain.

**Comment by makc** [comment]
I can imagine them going up at night, racing against the time to see the sunrise, for example. and then on their way back service people would come and turn them on. or, they would need to fix some failure that somehow requires to go higher. or, there would be long lines at the bottom.

**Comment by makc** [comment]
someone dropped this video in 3js discord https://youtu.be/kEB11PQ9Eo8

**Comment by felixmariotto** [comment]
> or, they would need to fix some failure that somehow requires to go higher.
Yes this sounds right.

Thanks for the video, it's super interesting ! We should do more of that kind of stuff with three.js... as the guy says, it's impossible to do this with Unity, so coding with something more bare like three would be an advantage (I suppose). And I love Portal.

**Comment by makc** [comment]
me this week:
![not really my game though](https://cdn.discordapp.com/attachments/217253743668756480/714028379573190716/Screenshot_2018-04-19-16-22-51.png)

but I am working towards this:
> If you have an idea for a nicer cave though go ahead

**Comment by felixmariotto** [comment]
> me this week:

😂 Was it un-lock-down were you live ? In France it was.

About the cave I was thinking, if you want the player to run for a while towards the camera, maybe you can programatically set some of those when entering your cave : 

https://github.com/felixmariotto/Edelweiss/blob/49931320695da62e2133238ac7c90dd8bdeadccf/public/js/CameraControl.js#L8-L12

In order to have the camera behave more adequately for backtracking ? I imagine increasing a bit the default distance, reducing the tweening speed, have a more vertical angle ?

**Comment by makc** [comment]
hey @felixmariotto I am finally at the point where I have everything I need for this cave, but I was thinking if there is a chance you remember how many stamina bars do we have when we first gets to this cave? If not, I would have to play all the way to there without cheats 😅 but I was thinking, we get +1 in the cave A, then we go through the cave B to get another +1 and that means 4 bars in total... right?

**Comment by felixmariotto** [comment]
Hi @makc, yes you have to get the first edelweiss in cave A, then after getting the dash ability you backtrack through cave B to get the second edelweiss with your new ability. At this point you have 4 stamina, then you go through the river and the cascade and you get the glider at the top of the forest biome. At this point you are supposed to backtrack again and get one of the two edelweisses left behind in the forest before to go through the canyon. The rest of the game can be done with 5 stamina, but there is still 3 edelweisses ahead.

**Comment by felixmariotto** [comment]
Which cave do you want to update ? All of them ?

**Comment by makc** [comment]
Idk, so far I am sort of just tying things out with this cave.

For example, diagonal limit planes:
![Screen Shot 2021-06-06 at 16 03 48 ](https://user-images.githubusercontent.com/242577/120927332-edd40080-c6e0-11eb-88cf-a1a019508ef5.png)
it was very easy to add, and I hoped this will be a good way to make a shortcut under the river, so that the player can run like this / instead of like this _| but in reality this makes very hard to design a wall behind the player, since it still has to be made of tiles that are not diagonal.

Another issue that I do not understand is why the camera is tilted to the right in this situation:
![Screen Shot 2021-06-06 at 16 04 25 ](https://user-images.githubusercontent.com/242577/120927571-9d10d780-c6e1-11eb-997d-496ddcc9b813.png)

**Comment by felixmariotto** [comment]
> Another issue that I do not understand is why the camera is tilted to the right in this situation:

About that, it's because I've made the camera check the distance of the nearest tiles on the right and left direction from the player, and make it turn toward the nearest tile. For instance if you have a tile very near at the right of the player, the camera will move to the left to show the tile more plainly. I have abolutely no idea where this bit of code is though lol

**Comment by makc** [comment]
> no idea where this bit of code is

regardless, in my case, if I add some dummy tiles on the right, the camera should straighten itself, correct?

**Comment by felixmariotto** [comment]
yes if the player is at equal distance between a tile on the right and one on the left, the camera should look straight ahead

**Comment by felixmariotto** [comment]
It's here : https://github.com/felixmariotto/Edelweiss/blob/4181238d5a50011309f7be714ffe60105d06dd9b/public/js/CameraControl.js#L129-L160

**Comment by makc** [comment]
It has only been 6 years, but I am this 🤏 close to finishing my cave 😅

<img width="777" height="392" alt="Image" src="https://github.com/user-attachments/assets/5d204577-7137-4435-8a92-f4c7132edc7c" />
